core/src/main/java/org/everit/json/schema/ObjectComparator.java

```diff
@@ -33,6 +45,15 @@ public static boolean deepEquals(Object obj1, Object obj2) {
                 return false;
             }
             return deepEqualObjects((JSONObject) obj1, (JSONObject) obj2);
+        } else if (obj1 instanceof Number) {
+            if (!(obj2 instanceof Number)) {
+                return false;
+            } else if (!obj1.getClass().equals(obj2.getClass())) {
+                String obj1AsString = obj1 instanceof Double && (Double) obj1 >= 10000000 ? numberFormat.format(obj1) : JSONObject.numberToString((Number) obj1);
+                String obj2AsString = obj2 instanceof Double && (Double) obj2 >= 10000000 ? numberFormat.format(obj2) : JSONObject.numberToString((Number) obj2);
```

Comment:
If the number is < 10000000 then it gets cast twice (once to Double, then to Number). Can we please avoid the duplicate casting? It will need an explicit if block.

Also the code is duplicated for `obj1` and `obj2` so please extract it to a `numberToString()` method or something similar.